pgmax/fg/graph.py

```diff
@@ -806,10 +910,23 @@ def BP(
         \tget_bp_state: Function to reconstruct the BPState from BPArrays.\n
         \tget_beliefs: Function to calculate beliefs from BPArrays.\n
     """
-    max_msg_size = int(np.max(bp_state.fg_state.wiring.edges_num_states))
-    num_val_configs = (
-        int(bp_state.fg_state.wiring.factor_configs_edge_states[-1, 0]) + 1
-    )
+    wiring = bp_state.fg_state.wiring
+    max_msg_size = int(np.max(wiring.edges_num_states))
+
+    num_val_configs = 0
+    if "EnumerationFactor" in wiring.wiring_by_factor_subtype:
```

Comment:
No special processing. Add more abstractions and rely on generic structures instead.

In this case we need to pre-compute `num_val_configs` because it's a static argument so we need its concrete value. For some factors we need this type of concrete arguments. For others we don't. This is an abstraction we should make: for each factor type we should have a function that return the necessary concrete arguments.

One convenient place to put this is as a property `static_arguments` of the corresponding wiring class, which in this case would be a dictionary `{'num_val_configs': self.factor_configs_edge_states[-1, 0]}`.

And in this section what we would do is
```
static_arguments = {
	factor_type: wiring[factor_type].static_arguments
	for factor_type in wiring
}
```